File: crates/brainwires-a2a/src/jsonrpc.rs
//! JSON-RPC 2.0 envelope types and method constants.

use serde::{Deserialize, Serialize};

use crate::error::A2aError;

// ---------------------------------------------------------------------------
// Method constants
// ---------------------------------------------------------------------------

/// Send a message to an agent.
pub const METHOD_MESSAGE_SEND: &str = "SendMessage";
/// Stream a message to an agent.
pub const METHOD_MESSAGE_STREAM: &str = "SendStreamingMessage";
/// Get a task by ID.
pub const METHOD_TASKS_GET: &str = "GetTask";
/// Cancel a task.
pub const METHOD_TASKS_CANCEL: &str = "CancelTask";
/// Resubscribe to task updates.
pub const METHOD_TASKS_RESUBSCRIBE: &str = "SubscribeToTask";
/// List tasks.
pub const METHOD_TASKS_LIST: &str = "ListTasks";
/// Set push notification configuration.
pub const METHOD_PUSH_CONFIG_SET: &str = "CreateTaskPushNotificationConfig";
/// Get push notification configuration.
pub const METHOD_PUSH_CONFIG_GET: &str = "GetTaskPushNotificationConfig";
/// List push notification configurations.
pub const METHOD_PUSH_CONFIG_LIST: &str = "ListTaskPushNotificationConfigs";
/// Delete push notification configuration.
pub const METHOD_PUSH_CONFIG_DELETE: &str = "DeleteTaskPushNotificationConfig";
/// Get authenticated extended agent card.
pub const METHOD_EXTENDED_CARD: &str = "GetExtendedAgentCard";

// ---------------------------------------------------------------------------
// Request ID
// ---------------------------------------------------------------------------

/// JSON-RPC request identifier (string or number).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RequestId {
    /// String identifier.
    String(String),
    /// Numeric identifier.
    Number(i64),
}

// ---------------------------------------------------------------------------
// JSON-RPC envelope types
// ---------------------------------------------------------------------------

/// JSON-RPC 2.0 request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcRequest {
    /// Protocol version (always "2.0").
    pub jsonrpc: String,
    /// Method name.
    pub method: String,
    /// Request parameters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<serde_json::Value>,
    /// Request identifier.
    pub id: RequestId,
}

/// JSON-RPC 2.0 response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcResponse {
    /// Protocol version (always "2.0").
    pub jsonrpc: String,
    /// Result on success.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    /// Error on failure.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<A2aError>,
    /// Request identifier echoed back.
    pub id: RequestId,
}

impl JsonRpcResponse {
    /// Create a success response.
    pub fn success(id: RequestId, result: serde_json::Value) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            result: Some(result),
            error: None,
            id,
        }
    }

    /// Create an error response.
    pub fn error(id: RequestId, error: A2aError) -> Self {
        Self {
            jsonrpc: "2.0".to_string(),
            result: None,
            error: Some(error),
            id,
        }
    }
}

File: deno/packages/a2a/jsonrpc.ts
/**
 * JSON-RPC 2.0 envelope types and A2A method constants.
 *
 * These are separate from MCP JSON-RPC types since A2A has its own
 * method namespace and error semantics.
 */

import type { A2aError } from "./error.ts";

// ---------------------------------------------------------------------------
// Method constants (A2A v1.0 PascalCase names)
// ---------------------------------------------------------------------------

/** Send a message to an agent. */
export const METHOD_MESSAGE_SEND = "SendMessage";
/** Stream a message to an agent. */
export const METHOD_MESSAGE_STREAM = "SendStreamingMessage";
/** Get a task by ID. */
export const METHOD_TASKS_GET = "GetTask";
/** Cancel a task. */
export const METHOD_TASKS_CANCEL = "CancelTask";
/** Resubscribe to task updates. */
export const METHOD_TASKS_RESUBSCRIBE = "SubscribeToTask";
/** List tasks. */
export const METHOD_TASKS_LIST = "ListTasks";
/** Set push notification configuration. */
export const METHOD_PUSH_CONFIG_SET = "CreateTaskPushNotificationConfig";
/** Get push notification configuration. */
export const METHOD_PUSH_CONFIG_GET = "GetTaskPushNotificationConfig";
/** List push notification configurations. */
export const METHOD_PUSH_CONFIG_LIST = "ListTaskPushNotificationConfigs";
/** Delete push notification configuration. */
export const METHOD_PUSH_CONFIG_DELETE = "DeleteTaskPushNotificationConfig";
/** Get authenticated extended agent card. */
export const METHOD_EXTENDED_CARD = "GetExtendedAgentCard";

// ---------------------------------------------------------------------------
// Request ID
// ---------------------------------------------------------------------------

/** JSON-RPC request identifier (string or number). */
export type RequestId = string | number;

// ---------------------------------------------------------------------------
// JSON-RPC envelope types
// ---------------------------------------------------------------------------

/** JSON-RPC 2.0 request. */
export interface JsonRpcRequest {
  /** Protocol version (always "2.0"). */
  jsonrpc: "2.0";
  /** Method name. */
  method: string;
  /** Request parameters. */
  params?: unknown;
  /** Request identifier. */
  id: RequestId;
}

/** JSON-RPC 2.0 response. */
export interface JsonRpcResponse {
  /** Protocol version (always "2.0"). */
  jsonrpc: "2.0";
  /** Result on success. */
  result?: unknown;
  /** Error on failure. */
  error?: A2aError;
  /** Request identifier echoed back. */
  id: RequestId;
}

/** Create a success response. */
export function createJsonRpcSuccess(
  id: RequestId,
  result: unknown,
): JsonRpcResponse {
  return { jsonrpc: "2.0", result, id };
}

/** Create an error response. */
export function createJsonRpcError(
  id: RequestId,
  error: A2aError,
): JsonRpcResponse {
  return { jsonrpc: "2.0", error, id };
}
